Localization/ensemble_kalman_filter/ensemble_kalman_filter.py

```diff
@@ -0,0 +1,229 @@
+NP = 20  # Number of Particle
+
+show_animation = True
+
+
+def calc_input():
+    v = 1.0  # [m/s]
+    yawrate = 0.1  # [rad/s]
+    u = np.array([[v, yawrate]]).T
+    return u
+
+
+
+def observation(xTrue, xd, u, RFID):
+
+    xTrue = motion_model(xTrue, u)
+
+    z = np.zeros((0, 4))
+
+    for i in range(len(RFID[:, 0])):
+
+        dx = RFID[i, 0] - xTrue[0, 0]
+        dy = RFID[i, 1] - xTrue[1, 0]
+        d = math.sqrt(dx**2 + dy**2)
+        angle = pi_2_pi(math.atan2(dy, dx) - xTrue[2, 0])
+        if d <= MAX_RANGE:
+            dn = d + np.random.randn() * Qsim[0, 0]  # add noise
+            anglen = angle + np.random.randn() * Qsim[1, 1]  # add noise
+            zi = np.array([dn, anglen,RFID[i, 0], RFID[i, 1]])
+            z = np.vstack((z, zi))
+
+    # add noise to input
+    ud = np.array([[
+        u[0, 0] + np.random.randn() * Rsim[0, 0],
+        u[1, 0] + np.random.randn() * Rsim[1, 1]]]).T
+
+    xd = motion_model(xd, ud)
+    return xTrue, z, xd, ud
+
+
+def motion_model(x, u):
+    F = np.array([[1.0, 0, 0, 0],
+                  [0, 1.0, 0, 0],
+                  [0, 0, 1.0, 0],
+                  [0, 0, 0, 0]])
+
+    B = np.array([[DT * math.cos(x[2, 0]), 0],
+                  [DT * math.sin(x[2, 0]), 0],
+                  [0.0, DT],
+                  [1.0, 0.0]])
+    x = F.dot(x) + B.dot(u)
+
+    return x
+
+
+def calc_LM_Pos(x, landmarks):
+    landmarks_pos=np.zeros((2*landmarks.shape[0],1))
+    for (i,lm) in enumerate(landmarks):
+        landmarks_pos[2*i] = x[0, 0] + lm[0] * math.cos(x[2, 0] + lm[1]) + np.random.randn() * Qsim[0, 0]/np.sqrt(2)
+        landmarks_pos[2*i+1] = x[1, 0] + lm[0] * math.sin(x[2, 0] + lm[1]) + np.random.randn() * Qsim[0, 0]/np.sqrt(2)
+    return landmarks_pos
+
+def calc_covariance(xEst, px):
+    cov = np.zeros((3, 3))
+
+    for i in range(px.shape[1]):
+        dx = (px[:, i] - xEst)[0:3]
+        cov +=  dx.dot(dx.T)
+
+    return cov
+
+
+def enkf_localization(px, xEst, PEst, z, u):
+    """
+    Localization with Ensemble Kalman filter
+    """
+    pz = np.zeros((z.shape[0]*2, NP))  # Particle store of z
+    for ip in range(NP):
+        x = np.array([px[:, ip]]).T
+
+        #  Predict with random input sampling
+        ud1 = u[0, 0] + np.random.randn() * Rsim[0, 0]
+        ud2 = u[1, 0] + np.random.randn() * Rsim[1, 1]
+        ud = np.array([[ud1, ud2]]).T
+        x = motion_model(x, ud)
+        px[:, ip] = x[:, 0]
+        z_pos=calc_LM_Pos(x, z)
+        pz[:, ip] = z_pos[:, 0]
+
+    x_ave=np.mean(px, axis=1)
+    x_dif=px - np.tile(x_ave,(NP,1)).T
+
+    z_ave=np.mean(pz, axis=1)
+    z_dif=pz - np.tile(z_ave,(NP,1)).T
+
+    U = 1/(NP-1)* x_dif @ z_dif.T
+    V = 1/(NP-1)* z_dif @ z_dif.T 
+
+    K = U @ np.linalg.inv(V) # Kalman Gain
+
+    z_lm_pos = z[:,[2,3]].reshape(-1,)
+
+    px_hat=px + K @ (np.tile(z_lm_pos,(NP,1)).T- pz)
+
+    xEst=np.average(px_hat, axis=1).reshape(4,1)
+    pEst=calc_covariance(xEst, px_hat)
```

Comment:
Why this is pEst? not PEst?

Reply:
Oh,I made a mistake.It should be PEst.Thank you.